homeassistant/components/trafikverket_train/sensor.py

```diff
@@ -0,0 +1,143 @@
+        vol.Required(CONF_FROM): cv.string,
+        vol.Optional(CONF_TIME): cv.time}]
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(
+        hass, config, async_add_devices, discovery_info=None):
+    """Set up the departure sensor."""
+    _LOGGER.debug("Start async_setup_platform")
+    from pytrafikverket import TrafikverketTrain
+    httpsession = async_get_clientsession(hass)
+    train_api = TrafikverketTrain(httpsession, config.get(CONF_API_KEY))
+    sensors = []
+    for train in config.get(CONF_TRAINS):
+        from_sig = yield from train_api.async_get_train_station(
+            train.get(CONF_FROM))
+        to_sig = yield from train_api.async_get_train_station(
+            train.get(CONF_TO))
+        sensor = TrainSensor(train_api,
+                             train.get(CONF_NAME),
+                             from_sig,
+                             to_sig,
+                             train.get(CONF_TIME))
+        sensors.append(sensor)
+
+    async_add_devices(sensors, update_before_add=True)
+    _LOGGER.debug("end async_setup_platform")
+
+
+class TrainSensor(Entity):
+    """Contains data about a train depature."""
+
+    def __init__(self, train_api, name, from_sig, to_sig, time):
+        """Initialize the sensor."""
+        self._train_api = train_api
+        self._name = name
+        self._from_sig = from_sig
+        self._to_sig = to_sig
+        self._time = time
+        self._state = None
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Retrieve latest state."""
+        _LOGGER.debug("start async_update")
+        if self._time is not None:
+            when = datetime.combine(date.today(), self._time)
+            _LOGGER.debug("self._from_sig: %s, self._to_sig: %s, when: %s",
+                          self._from_sig.name, self._to_sig.name, when)
+            self._state = yield from \
+                self._train_api.async_get_train_stop(
+                    self._from_sig, self._to_sig, when)
+        else:
+            when = datetime.now()
+            _LOGGER.debug("self._from_sig: %s, self._to_sig: %s, when: %s",
+                          self._from_sig.name, self._to_sig.name, when)
+            self._state = yield from \
+                self._train_api.async_get_next_train_stop(
+                    self._from_sig, self._to_sig, when)
+        _LOGGER.debug("end async_update")
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        if self._state is None:
+            return None
+        state = self._state
+        other_information = None
+        if state.other_information is not None and state.other_information:
+            other_information = ", ".join(state.other_information)
+        deviations = None
+        if state.deviations is not None and state.deviations:
+            deviations = ", ".join(state.deviations)
+        delay_in_minutes = state.get_delay_time()
```

Comment:
@ljmerza Nope, this will not render an API call. 
I calls a method in pytrafikverket:

```
    def get_delay_time(self) -> timedelta:
        if self.canceled:
            return None
        if (self.advertised_time_at_location is not None and
                self.time_at_location is not None and
                self.advertised_time_at_location != self.time_at_location):
            return self.time_at_location - self.advertised_time_at_location
        if (self.advertised_time_at_location is not None and
                self.estimated_time_at_location is not None and
                self.advertised_time_at_location != self.estimated_time_at_location):
            return self.estimated_time_at_location - self.advertised_time_at_location
        return None
```

Reply:
Moved the method calls to async_update() to have it cleaner. Even if there was no API call involved.